### PR TITLE
Added the weather control device.

### PR DESCRIPTION
**Pull request by Mailaender**
Closes https://github.com/OpenRA/ra2/issues/149.

### CODE REVIEW

**Comment by Mailaender** [comment]
Try changing OffsetsX/Y and WeaponCount and the LightningStorm SpreadDamange to see if you can adjust it to match what you expect.

**Comment by sorcerer86pt** [comment]
ok, will do

**Comment by Mailaender** [comment]
Added the thunder sounds thanks to @GraionDilach.

**Comment by Mailaender** [comment]
Simplified the code and added more damage.

**Comment by sorcerer86pt** [comment]
Tried to add weathercloud4 and 5, and also the respective bolts in misc.yml, under the explosion part. Also added that part to the zap.yml

```
diff --git a/mods/ra2/sequences/misc.yaml b/mods/ra2/sequences/misc.yaml
index 6ac3135..1cac440 100644
--- a/mods/ra2/sequences/misc.yaml
+++ b/mods/ra2/sequences/misc.yaml
@@ -231,6 +231,18 @@ explosion:
        weatherbolt3: wclbolt3
                Offset: -0, 80
                Tick: 200
+  weathercloud4: wccloud4
+    Offset: 0, -120
+    Tick: 80
+  weatherbolt4: wclbolt4
+    Offset: 0, 60
+    Tick: 200
+  weathercloud5: wccloud5
+    Offset: -0, -100
+    Tick: 80
+  weatherbolt5: wclbolt5
+    Offset: -0, 80
+    Tick: 200

 chrono:
        warpin: warpin
diff --git a/mods/ra2/weapons/zaps.yaml b/mods/ra2/weapons/zaps.yaml
index 1bbb852..76c3429 100644
--- a/mods/ra2/weapons/zaps.yaml
+++ b/mods/ra2/weapons/zaps.yaml
@@ -238,7 +238,7 @@ LightningStorm:
                Blockable: false
        Warhead@1Eff: CreateEffect
                Delay: 0
-               Explosions: weathercloud1, weathercloud2, weathercloud3
+               Explosions: weathercloud1, weathercloud2, weathercloud3, weathercloud4, weathercloud5
                UsePlayerPalette: true
                ExplosionPalette: player
                ValidTargets: Ground, Water
@@ -262,7 +262,7 @@ LightningStorm:
                DamageTypes: Prone100Percent
        Warhead@3Eff: CreateEffect
                Delay: 50
-               Explosions: weatherbolt1, weatherbolt2, weatherbolt3
+               Explosions: weatherbolt1, weatherbolt2, weatherbolt3, weatherbolt4, weatherbolt5
                ValidTargets: Ground, Water
                ImpactSounds: sweastra.wav, sweastrb.wav, sweastrc.wav, sweastrd.wav
        Warhead@4Smu: LeaveSmudge
```

But when i deploy the weather storm, the game crashs with: 

```
[2020-03-14T19:36:10] Game started
Exception of type `System.InvalidOperationException`: Unit `explosion` does not have a sequence named `weathercloud4`
   em OpenRA.Graphics.SequenceProvider.GetSequence(String unitName, String sequenceName)
   em OpenRA.Graphics.Animation.PlaySequence(String sequenceName)
   em OpenRA.Graphics.Animation.PlayThen(String sequenceName, Action after)
   em OpenRA.Mods.Common.Effects.SpriteEffect..ctor(Func`1 posFunc, Func`1 facingFunc, World world, String image, String sequence, String palette, Boolean visibleThroughFog, Boolean scaleSizeWithZoom)
   em OpenRA.Mods.Common.Warheads.CreateEffectWarhead.<>c__DisplayClass12_0.<DoImpact>b__0(World w)
   em OpenRA.World.Tick()
   em OpenRA.Game.InnerLogicTick(OrderManager orderManager)
   em OpenRA.Game.LogicTick()
   em OpenRA.Game.Loop()
   em OpenRA.Game.Run()
   em OpenRA.Game.InitializeAndRun(String[] args)
   em OpenRA.Program.Main(String[] args)

```

What i'm missing?

**Comment by Mailaender** [comment]
Those sprites don't seem to exit.

**Comment by sorcerer86pt** [comment]
how i add them?

**Comment by Mailaender** [comment]
They aren't in the game files. That's why it is crashing.

**Comment by Mailaender** [comment]
Corrected the offsets again thanks to @GraionDilach and made the duration much longer.

**Comment by sorcerer86pt** [comment]
ok, since you corrected the offsets and duration i will check that out

**Comment by sorcerer86pt** [comment]
it only works once. And the offsets are strange, I will put a movie of that in play

https://drive.google.com/file/d/1xQKwNSGzGpxNBV-JzmMIvOpMyGVqqCtg/view?usp=sharing

**Comment by Mailaender** [comment]
Fixed the regression where it only fired once. Due to the balancing: I assume tanks act as a faraday cage. It also does little damage to the construction yard which I guess is intended.

**Comment by sorcerer86pt** [comment]
yes, the con yard and some other buildings (like support weapons and super weapons) should survive 1 super weapon.

**Comment by sorcerer86pt** [comment]
still only fire once.

**Comment by Mailaender** [comment]
Updated again with https://github.com/GraionDilach/ra2/commit/99d0a6d74e8b34a7df55724e8c01206d6e594793 to match the original even further.

**Comment by sorcerer86pt** [comment]
ok, it's working for several launches of the storm, and the animation looks ok, 2 things now:

- Recheck the damage values against armor, those 0 in the zap.yml. I think it's the reason why it doesn't affect units and only buildings,
- Check if you know a way to just give 1 warning of the weather storm. I heard it multiple times during a single launch

File: mods/ra2/weapons/zaps.yaml [hunk 1]
```diff
@@ -230,3 +230,45 @@ CRNeutronRifle:
...
+			Flak: 0
+			Plate: 0
+			Light: 0
+			Medium: 0
+			Heavy: 0
+			Wood: 100
+			Steel: 100
+			Concrete: 6
+			Drone: 0
+			Rocket: 0
```
Comment (sorcerer86pt) [inline mods/ra2/weapons/zaps.yaml]: Those values are okay? Those 0 there

Comment (sorcerer86pt) [reply mods/ra2/weapons/zaps.yaml]: Changed the values in this part, and it started to affect units.

Comment (sorcerer86pt) [reply mods/ra2/weapons/zaps.yaml]: btw the OG ra2 yuris revenge verses are:

```
; Ion storm strike ;lightning strike
[IonWH]
Wall=yes
Wood=yes
CellSpread=2.0
PercentAtMax=.5
Verses=100%,100%,100%,100%,100%,100%,100%,100%,2%,100%,100%
;;Verses=100%,100%,100%,100%,100%,100%,100%,100%,3%,100%,100%
Conventional=yes
Rocker=yes
InfDeath=5
;AnimList=XGRYSML1,XGRYSML2,EXPLOSML,XGRYMED1,XGRYMED2,EXPLOMED,EXPLOLRG
AnimList=EXPLOSML
;Deform=10%
;DeformThreshhold=300
;Tiberium=yes
;Sparky=no
Bright=yes
CombatLightSize=40%	; SJM: Force size of combat light, overriding damage-based value
```

**Comment by sorcerer86pt** [comment]
Found another issue. A lighting storm against a weather controller device doesn't do any damage. 
To test:

We will have to enable some debug options, namely 

- build everywhere,
- instant charge time
- build everything
- and give some money

1. Create a weather device in your base.
2. Then create another one somewhere on map, far away from base
3. Create another building there without being the weather controller (for example barracks)
4. Cast Lighting storm on the far away Weather Controller device

File: mods/ra2/weapons/zaps.yaml (hunk 1 above)
Comment (Mailaender) [reply mods/ra2/weapons/zaps.yaml]: Made it more deadly.

**Comment by Mailaender** [comment]
> Check if you know a way to just give 1 warning of the weather storm. I heard it multiple times during a single launch

I guess you have cheats on and hear "Lightning Storm created" and "Lightning Storm ready" at the same time.

> A lighting storm against a weather controller device doesn't do any damage.

That isn't a bug. That is balancing on purpose. I assume they wanted people to get forced to attack with ground units against a super weapon or construction yard instead of turtle in your base and build superweapons against superweapons.

**Comment by sorcerer86pt** [comment]
I was trying to test the damage of weather storm against various buildings
and units to balance the versus damage. So of course I build the nuke
launcher, chronosphere, iron curtain and weather controller and dropped the
storm there.



A domingo, 15/03/2020, 7:45 da manhã, Matthias Mailänder <
notifications@github.com> escreveu:

**Comment by Mailaender** [comment]
That is because it has `Armor: Concrete`. It's not a bug.

**Comment by sorcerer86pt** [comment]
Nuke launcher doesn't also has it? Because it was more damaged than weather
controller device.


Em 15/03/2020 à(s) 11:27 da manhã, Matthias Mailänder <
notifications@github.com> escreveu:

That is because it has Armor: Concrete. It's not a bug.

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub
<https://github.com/OpenRA/ra2/pull/655#issuecomment-599197838>, or
unsubscribe

**Comment by Mailaender** [comment]
Check the original Rules.ini if the entries don't match.

**Comment by sorcerer86pt** [comment]
well both buildings have the concrete armor in their respective yaml files.

**Comment by sorcerer86pt** [comment]
I tried the following damage and versus configuration, and it looked a good start.

```
Warhead@2Dam: SpreadDamage
		Delay: 56
		Spread: 3c0
		Damage: 100
		Versus:
			None: 100
			Flak: 100
			Plate: 100
			Light: 100
			Medium: 95
			Heavy: 90
			Wood: 70
			Steel: 50
			Concrete: 25
			Drone: 100
			Rocket: 100
```

Reduced the damage each bolt does individually, but since it has a great deal of them, it spreads the damage.,

**Comment by MustaphaTR** [comment]
> I was trying to test the damage of weather storm against various buildings
and units to balance the versus damage. So of course I build the nuke
launcher, chronosphere, iron curtain and weather controller and dropped the
storm there. All other 3 buildings were damaged, only the weather device not.

This sounds like because of lack of `AffectsParent: true`.

Another thing that can cause a different damage output than original is the fact that buildings took damage for each cell of theirs for warheads with bigger than 1 cell spread. They only take that damage once here.

**Comment by sorcerer86pt** [comment]
I think it's that flag. Other buildings damage ( nuke silo) takes damage as
expected.


Em 15/03/2020 à(s) 3:47 da tarde, Mustafa Alperen Seki <
notifications@github.com> escreveu:

I was trying to test the damage of weather storm against various buildings
and units to balance the versus damage. So of course I build the nuke
launcher, chronosphere, iron curtain and weather controller and dropped the
storm there. All other 3 buildings were damaged, only the weather device
not.

This sounds like because of lack of AffectsParent: true.

Another thing that can cause a different damage output than original is the
fact that buildings took damage for each cell of theirs for warheads with
bigger than 1 cell spread. They only take that damage once here.

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub
<https://github.com/OpenRA/ra2/pull/655#issuecomment-599227426>, or
unsubscribe

**Comment by Mailaender** [comment]
Added `AffectsParent = true` so that the weather control device can damage itself.

**Comment by sorcerer86pt** [comment]
ok, it affects now the weather controller device. I'm okay with merging this as it is, IF we open a Enhancement for balancing the damage. 2% is too bad for this, seeing that the nuke does effective about 49% damage (that is a nuke targeted directly at the weather controller device damage it at the limit to still have the green bar)

File: mods/ra2/weapons/zaps.yaml [hunk 2]
```diff
@@ -230,3 +230,46 @@ CRNeutronRifle:
...
+		Versus:
+			None: 100
+			Flak: 100
+			Plate: 100
+			Light: 100
+			Medium: 100
+			Heavy: 100
+			Wood: 100
+			Steel: 100
+			Concrete: 2
```
Comment (MustaphaTR) [inline mods/ra2/weapons/zaps.yaml]: This is 3% in RA2, 2% in YR. Not a big difference but this mod uses RA2 values everywhere else.

Comment (Mailaender) [reply mods/ra2/weapons/zaps.yaml]: Changed.

**Pull request merged**